packages/babel-plugin-proposal-object-rest-spread/src/index.js

```diff
@@ -209,12 +214,76 @@ export default declare((api, opts) => {
     inherits: syntaxObjectRestSpread,
 
     visitor: {
-      // taken from transform-parameters/src/destructuring.js
       // function a({ b, ...c }) {}
       Function(path) {
         const params = path.get("params");
-        for (let i = params.length - 1; i >= 0; i--) {
-          replaceRestElement(params[i].parentPath, params[i]);
+        const paramsWithRestElement = new Set();
+        const idsInRestParams = new Set();
+        for (let i = 0; i < params.length; ++i) {
+          const param = params[i];
+          if (hasRestElement(param)) {
+            paramsWithRestElement.add(i);
+            for (const name of Object.keys(param.getBindingIdentifiers())) {
+              idsInRestParams.add(name);
+            }
+          }
+        }
+
+        // if true, a parameter exists that has an id in its initializer
+        // that is also an id bound in a rest parameter
+        // example: f({...R}, a = R)
+        let idInRest = false;
+
+        const IdentifierHandler = function(idPath) {
+          if (
+            (idPath.parentKey === "right" || idPath.parentKey === "value") &&
```

Comment:
It does not suffice to check `parentKey` to determine if `idPath` is _referenced_. i.e.
```js
({ ...R }, a = f(R)) => {}
({ ...R }, { [R.key]: a }) => {}
```

We can check if an identifier is referencing the rest param binding via `scope`, like what we did in
https://github.com/babel/babel/blob/548cb3ee89552ffc08ee5625b084bd33f8107530/packages/babel-plugin-transform-parameters/src/params.js#L26
```js
{
  "ReferencedIdentifier|BindingIdentifier"(path, state) {
      // state.scope is the function scope
      // path.scope is the scope of current traversed identifier
      scope.getBinding(name) === state.scope.getBinding(name) && // referencing the binding accessible from function scope
      state.scope.hasOwnBinding(name) && // the binding is owned by the function scope
      idsInRestParams.has(name) // it a binding introduced by the rest param
    }
}
```

And similar to #11349, if we are doing on this approach, we shall skip type annotations and type parameters.

Reply:
Fixed. I swear every time I think I need to do something manually, Babel's path/visitor api proves me wrong and gives me a shorter solution.